lib/pbench/server/api/resources/metadata_api.py

```diff
@@ -0,0 +1,355 @@
+                        "updated": "datetime string",
+                        "key": favorite/saved
+                        }
+                }
+        """
+        metadata_key = key.upper()
+        if metadata_key not in [key.name for key in MetadataKeys]:
+            self.logger.warning(
+                "Invalid Metadata key provided during metadata creation. Key: {}",
+                metadata_key,
+            )
+            abort(
+                400,
+                message="Invalid metadata key. Key need to be either Favorite/Saved",
+            )
+
+        post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        current_user = self.auth.token_auth.current_user()
+        if current_user:
+            current_user_id = current_user.id
+        else:
+            current_user_id = None
+
+        value = post_data.get("value")
+        if not value:
+            self.logger.warning(
+                "value not provided during metadata creation. user_id: {}",
+                current_user_id,
+            )
+            abort(400, message="Value field missing")
+
+        try:
+            # Create a new metadata session
+            metadata_session = Metadata(
+                value=value, key=metadata_key, user_id=current_user_id
+            )
+            # insert the metadata session for a user
+            metadata_session.add()
+            self.logger.info(
+                "New metadata session created for user_id {}", current_user_id
+            )
+        except Exception:
+            self.logger.exception("Exception occurred during the Metadata creation")
+            abort(500, message="INTERNAL ERROR")
+        else:
+            response_object = {
+                "message": "success",
+                "data": metadata_session.get_json(),
+            }
+            return make_response(jsonify(response_object), 201)
+
+    @Auth.token_auth.login_required(optional=True)
+    def get(self, key):
+        """
+        Get request for querying all the metadata sessions for a user.
+        returns the list of all the metadata sessions associated with a logged in user.
+        This requires a Pbench auth token in the header field
+
+        Required headers include
+            Authorization:   Bearer <Pbench authentication token (user received upon login)>
+
+        :return: JSON Payload
+            response_object = {
+                    "message": "success"
+                    "data": {
+                        "sessions": [
+                            {"id": "metadata_id",
+                            "value": "client text blob",
+                            "key": "key string",
+                            "created": "datetime string",
+                            "updated": "datetime string", },
+                            {}]
```

Comment:
Yeah I was having a discussion around this point the other day with Dave, Its definitely possible to just return a list of IDs and then the client can query the specific information about the an object using an ID in the list. However, most of the time the client might end up asking more information about all the metadata IDs since nothing can be inferred from just the IDs they are auto-incremented primary keys in our database. 
and yes I will update the docstring

Reply:
I think I need a better understanding of the work-flow that this is intended to support.

In particular, why would the client want to fetch all of the IDs/values, and how would doing so help?  The client really should know what it wants, and I'm not sure how it can infer more from the data than it can from just the IDs (that is, isn't the data just key-value pairs?...so having multiple instances of those (which probably all have the same keys) doesn't actually resolve anything...right?).

At this point, I'm thinking that this API isn't actually useful (for anything other than, say, server maintenance or debugging).  Do we really have a requirement for this?

Reply:
The dashboard currently fetches "all saved view" data, which corresponds to the list of metadata with whatever key we assign for saved views (e.g., in this PR, "saved"). This would be a list of saved='text' pairs, with each value representing the (user provided) description and redux context of a particular dashboard view saved by the user. The dashboard can then display a table containing this information to help the user choose one. (Look at the Sessions tab on http://pbench.perf.lab.eng.bos.redhat.com/dashboard for example.)

It's interesting that (at least the last time I looked at the code) the dashboard is then re-fetching the specific "session id" (which in our model is the `user_metadata` SQL generated row ID) before restoring the redux context to dispatch to the saved view.

I can't imagine offhand any case for returning all keys, aside from "maintenance", but all values for a specific key seems like a pretty straightforward use case.

Reply:
> Look at the Sessions tab on http://pbench.perf.lab.eng.bos.redhat.com/dashboard for example.
I don't see a "Sessions" tab -- I see a "Controllers" tab and a "Favorites" tab -- perhaps because I haven't created any Sessions?

> I can't imagine offhand any case for returning all keys, aside from "maintenance", but all values for a specific key seems like a pretty straightforward use case.

This is near where I'm getting hung up, I think.  We have a two-dimensional matrix -- a set of ID's and a set of keys -- and each cell in the matrix contains a value.  I can imagine (for a single user) wanting to slice this to produce a list of values, one for each key, for a single ID (this might be termed a "session context"), and I can imagine wanting a slice for a single key (and user), producing a list of IDs of what we might call "sessions".  So, the traversal goes like this:  we have a body of metadata, divided into some number of keys, each of which has some number of instances corresponding to an ID, each of which contains a value; so, one might want the list of keys, or, for a given key one might want the list of IDs (on a per-user basis), or, for a given ID, one might want the list of key/value pairs (for a given user).

I suppose that one could implement this by providing an API which returns the whole matrix and let the client paw through it, but it seems like we could do better, having the client request the slice that it wants and returning only that.

Reply:
> > Look at the Sessions tab on http://pbench.perf.lab.eng.bos.redhat.com/dashboard for example.
> 
> I don't see a "Sessions" tab -- I see a "Controllers" tab and a "Favorites" tab -- perhaps because I haven't created any Sessions?

OK, fine, "tab" is a versatile term. I'm talking about the separate views in the left sidebar: from top to bottom, Dashboard, Search, and Sessions. If your sidebar is closed, click on the hamburger at the top to open it.

> > I can't imagine offhand any case for returning all keys, aside from "maintenance", but all values for a specific key seems like a pretty straightforward use case.
> 
> This is near where I'm getting hung up, I think. We have a two-dimensional matrix -- a set of ID's and a set of keys -- and each cell in the matrix contains a value. I can imagine (for a single user) wanting to slice this to produce a list of values, one for each key, for a single ID (this might be termed a "session context"), and I can imagine wanting a slice for a single key (and user), producing a list of IDs of what we might call "sessions". So, the traversal goes like this: we have a body of metadata, divided into some number of keys, each of which has some number of instances corresponding to an ID, each of which contains a value; so, one might want the list of keys, or, for a given key one might want the list of IDs (on a per-user basis), or, for a given ID, one might want the list of key/value pairs (for a given user).
> 
> I suppose that one could implement this by providing an API which returns the whole matrix and having the client paw through it, but it seems like we could do better, having the client request the slice that it wants and returning only that.

As Nikhil has pointed out, it's not really a matrix. We're storing it all in one table, with a unique integer row ID. Both the key and value are columns; it doesn't make much sense to search by value (though we could), but it definitely makes sense to search by key so you see only the values you want for a particular purpose (e.g., saved views or favorites).

So we've really just got a vector of values uniquely identified by row ID but also with a finite set of key values that are handy to slice the vector into interesting chunks based on your intent. So your final "for a given ID [...] the list of key/value pairs" doesn't make any sense: a given ID by definition has a single key/value.

Reply:
Ah.  On my sidebar, I've got "Dashboard", "Search", and "_Explore_"...but, when I click on "Explore", I get a list of "Shared Sessions".

> a given ID by definition has a single key/value.

Ah, right.  That's true.

So, the client can say "for this ID that I remember creating, give me the back the value (and give me the key, too, because reasons?)", or the client can say "for this key, give me the values (and give me the corresponding IDs, so that I can update or delete the value of my choice later)".

I'm thinking that the first of those two is of very limited value.  That is, if the client can remember the ID, then it can remember the value; so, it's only useful if the client has list of IDs and it's iterating over them, and you guys think that we should just return the values with the IDs and cut out the intermediate step.

So, the workflow or use case then is the client starts with a key and asks for a list of values (and corresponding IDs) which this user has associated with that key.  So, it does a `GET /metadata/<key>` and back they come.  (Or, it does a `GET /metadata?key=<key>`...I'm happy either way.)